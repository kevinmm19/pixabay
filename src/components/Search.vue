<template>
  <section class="search">
      <form>
          <label for="searchInput" class="search__label">Search: </label>
          <input type="search" id="searchInput" class="search__input js-search" @input="fetch" placeholder="Type anything..." v-model="search" ref="searchInput" disabled="disabled">
      </form>
  </section>
</template>

<script>
export default {
  name: 'Search',
  data: function () {
    return {
        search: '',
        timeout: 0
    }
  },
  async mounted() {
    await this.find();
    this.$refs.searchInput.removeAttribute('disabled');
    this.debounceFind = this.debounce(this.find, 250);
  },
  destroyed() {
    clearTimeout(this.timeout);
  },
  methods: {
    debounce(func, wait, immediate) {
        return function() {
            let context = this, args = arguments;
            let later = function() {
            this.timeout = null;
            if (!immediate) func.apply(context, args);
            };
            let callNow = immediate && !this.timeout;
            clearTimeout(this.timeout);
            this.timeout = setTimeout(later, wait);
            if (callNow) func.apply(context, args);
        };
    },
    async find() {
        try {
            let error429 = false;
            const query = this.search.trim().replace(/[|&;$%@"<>()+=,]/g, '');
            const url = 'https://pixabay.com/api/?key=9656065-a4094594c34f9ac14c7fc4c39&per_page=6&q=' + query;
            const headers = new Headers({
                'Content-Type': 'text/plain'
            });
            const response = await fetch(url, { 
                method: 'GET',
                mode: 'cors',
                headers: headers
            });
            if (response !== undefined && response.ok) {
                const images = await response.json();
                this.$store.commit({
                  type: 'setImages',
                  list: images.hits
                });
            } else {
                error429 = true;
            }
            this.$store.commit({
              type: 'setFlags',
              errorFetch: false,
              error429: error429,
              loading: false
            });
        } catch (error) {
            this.$store.commit({
              type: 'setFlags',
              errorFetch: true,
              error429: false,
              loading: false
            });
        }
    },
    fetch() {
        this.$store.commit({
          type: 'setFlags',
          errorFetch: false,
          error429: false,
          loading: true
        });
        this.debounceFind();
    }
  }
}
</script>

<style lang="scss" scoped>
@import "../scss/_vars.scss";
@import "../scss/_mixins.scss";
.search {
    position: fixed;
    padding: 20px 0;
    width: 100%;
    text-align: center;
    background-color: $colorBg1;
    z-index: 1;
    &__label {
        color: $colorBg5;
        font-size: 18px;
        padding-right: 5px;
    }
    &__input {
        padding: 10px;
        height: 20px;
        width: 68%;
        @include respond-to($tablet) {
            width: 50%;
        }
    }
}
</style>
