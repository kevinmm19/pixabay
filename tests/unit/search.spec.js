import Vue from 'vue';
import Search from "@/components/Search.vue";

describe("Search.vue", () => {
  it('sets the correct default data', () => {
    expect(typeof Search.data).toBe('function');
    const defaultData = Search.data();
    expect(defaultData.search).toBe('');
  });
  it("renders correctly", done => {
    const vm = new Vue(Search).$mount();
    vm.$nextTick(() => {
      expect(vm.$el).toMatchSnapshot();
      done();
    });
  });
  it("renders correctly after changing the search state", done => {
    const vm = new Vue(Search).$mount();
    vm.search = 'roses';
    vm.$nextTick(() => {
      expect(vm.$el).toMatchSnapshot();
      done();
    });
  });
});
