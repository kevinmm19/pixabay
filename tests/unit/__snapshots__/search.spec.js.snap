// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search.vue renders correctly 1`] = `
<section
  class="search"
>
  <form>
    <label
      class="search__label"
      for="searchInput"
    >
      Search: 
    </label>
     
    <input
      class="search__input js-search"
      disabled="disabled"
      id="searchInput"
      placeholder="Type anything..."
      type="search"
    />
  </form>
</section>
`;

exports[`Search.vue renders correctly after changing the search state 1`] = `
<section
  class="search"
>
  <form>
    <label
      class="search__label"
      for="searchInput"
    >
      Search: 
    </label>
     
    <input
      class="search__input js-search"
      disabled="disabled"
      id="searchInput"
      placeholder="Type anything..."
      type="search"
    />
  </form>
</section>
`;
